handlers.go

```diff
@@ -229,15 +229,54 @@ func (a *autographer) handleSignature(w http.ResponseWriter, r *http.Request) {
 	log.WithFields(log.Fields{"rid": rid}).Info("signing request completed successfully")
 }
 
-// handleHeartbeat returns a simple message indicating that the API is alive and well
-func handleHeartbeat(w http.ResponseWriter, r *http.Request) {
+// handleLBHeartbeat returns a simple message indicating that the API is alive and well
+func handleLBHeartbeat(w http.ResponseWriter, r *http.Request) {
 	if r.Method != "GET" {
 		httpError(w, r, http.StatusMethodNotAllowed, "%s method not allowed; endpoint accepts GET only", r.Method)
 		return
 	}
 	w.Write([]byte("ohai"))
 }
 
+// handleHeartbeat checks whether backing services are enabled and
+// accessible and returns 200 when they are and 502 when the
+// aren't. Currently it only checks whether the HSM is accessible.
+func (a *autographer) handleHeartbeat(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		httpError(w, r, http.StatusMethodNotAllowed, "%s method not allowed; endpoint accepts GET only", r.Method)
+		return
+	}
+	var (
+		// a map of backing service name to up or down/inaccessible status
+		result = map[string]bool{}
+		status = http.StatusOK
+	)
+
+	// for the first signer with a conf.PrivateKey value short
+	// enough to be an HSM label try to fetch it from the HSM
+	conf := a.hsmHeartbeatSignerConf
+	if conf != nil {
+		err := conf.CheckHSMConnection()
+		if err == nil {
+			result["hsmAccessible"] = true
+			status = http.StatusOK
+		} else {
+			log.Errorf("error checking HSM connection for signer %s: %s", conf.ID, err)
+			result["hsmAccessible"] = false
+			status = http.StatusBadGateway
```

Comment:
perhaps a 500 internal server error is more appropriate to return than a 502, so those errors can't be confused with the 502s nginx may return?